host/bcm283x/dma.go

```diff
@@ -655,6 +659,60 @@ func allocateCB(size int) ([]controlBlock, *videocore.Mem, error) {
 	return cb, buf, nil
 }
 
+// dmaWriteStreamPCM streams data to a PCM enabled pin as a half-duplex I²S
+// channel.
+func dmaWriteStreamPCM(p *Pin, w gpiostream.Stream) error {
+	if w.Duration() == 0 {
+		return nil
+	}
+	resolution := w.Resolution()
+	hz := uint64(time.Second / resolution)
+	_, _, _, actualHz, err := calcSource(hz, 1)
+	if err != nil {
+		return err
+	}
+	if actualHz != hz {
+		return errors.New("TODO(maruel): handle oversampling")
+	}
+
+	// Start clock earlier.
+	pcmMemory.reset()
+	_, _, err = setPCMClockSource(hz)
+	if err != nil {
+		return err
+	}
+
+	l := (int(w.Duration()/resolution) + 7) / 8 // Bytes
+	buf, err := dmaBufAllocator((l + 0xFFF) &^ 0xFFF)
+	if err != nil {
+		return err
+	}
+	defer buf.Close()
+	if err := copyStreamToDMABuf(w, buf.Uint32()); err != nil {
+		return err
+	}
+
+	cb, pCB, err := allocateCB(4096)
+	if err != nil {
+		return err
+	}
+	defer pCB.Close()
+	reg := pcmBaseAddr + 0x4 // pcmMap.fifo
+	if err := cb[0].initBlock(uint32(buf.PhysAddr()), reg, uint32(l), false, true, true, false, dmaPCMTX); err != nil {
+		return err
+	}
+
+	defer pcmMemory.reset()
+	// Start transfer
+	pcmMemory.set()
+	runIO(pCB, l <= maxLite)
+	// We have to wait PCM to be finished even after DMA finished.
+	for pcmMemory.cs&pcmTXErr == 0 {
+		Nanospin(10 * time.Nanosecond)
```

Comment:
In theory it should try to wait the exact right amount of time.

Reply:
Yeah, I just wanted avoid a busy loop. Maybe, we should wait 1 PCM clock.
dmaChannel.wait() is a busy loop now.